# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

## Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [2]:
# TODO: import the pandas library
import pandas as pd
import numpy as np

# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('../data/rural_population_percent.csv', header=2)
df_electricity = pd.read_csv('../data/electricity_access_percent.csv', header=2)

# TODO: remove the Unnamed: 62 column from each data set

# TODO: combine the two data sets together using the concat method
# In other words, all of the rows of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.

In [3]:
df_rural.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965  ...  \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300  ...   
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250  ...   
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496  ...   
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770  ...   
4  SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555  ...   

     2009    2010    2011    2012    2013    2014    2015    2016    2017  \
0  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696  58.893   
1  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868  72.430   
2  60.701  59.903  59.100  58.301  57.510  56.726  55.950  55.181  54.422   
3  48.924  47.837  46.753  45.670  44.617  43.591  42.593  41.624  40.684   
4  11.648  12.183  12.740  13.292  13.835  14.367  14.885  15.388  15.873   

   Unnamed: 62  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 63 columns]

In [4]:
df_electricity.head()

Country Name Country Code                           Indicator Name  \
0        Aruba          ABW  Access to electricity (% of population)   
1  Afghanistan          AFG  Access to electricity (% of population)   
2       Angola          AGO  Access to electricity (% of population)   
3      Albania          ALB  Access to electricity (% of population)   
4      Andorra          AND  Access to electricity (% of population)   

   Indicator Code  1960  1961  1962  1963  1964  1965  ...        2009  \
0  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   93.354546   
1  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   44.854885   
2  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   32.382469   
3  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  100.000000   
4  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  100.000000   

         2010        2011        2012        2013        2014        2015  \
0   93.356292   93.942375   94.255814   94.578262   94.906723   95.238182   
1   42.700000   43.222019   69.100000   67.259552   89.500000   71.500000   
2   33.514950   34.600000   35.821964   36.990490   32.000000   42.000000   
3  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   
4  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   

         2016  2017  Unnamed: 62  
0   95.570145   NaN          NaN  
1   84.137138   NaN          NaN  
2   40.520607   NaN          NaN  
3  100.000000   NaN          NaN  
4  100.000000   NaN          NaN  

[5 rows x 63 columns]

In [6]:
df_rural.drop('Unnamed: 62', axis=1, inplace=True)
df_electricity.drop('Unnamed: 62', axis=1, inplace=True)

In [9]:
pd.concat([df_rural,df_electricity])

Country Name Country Code                            Indicator Name  \
0           Aruba          ABW  Rural population (% of total population)   
1     Afghanistan          AFG  Rural population (% of total population)   
2          Angola          AGO  Rural population (% of total population)   
3         Albania          ALB  Rural population (% of total population)   
4         Andorra          AND  Rural population (% of total population)   
..            ...          ...                                       ...   
259        Kosovo          XKX   Access to electricity (% of population)   
260   Yemen, Rep.          YEM   Access to electricity (% of population)   
261  South Africa          ZAF   Access to electricity (% of population)   
262        Zambia          ZMB   Access to electricity (% of population)   
263      Zimbabwe          ZWE   Access to electricity (% of population)   

     Indicator Code    1960    1961    1962    1963    1964    1965  ...  \
0    SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300  ...   
1    SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250  ...   
2    SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496  ...   
3    SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770  ...   
4    SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555  ...   
..              ...     ...     ...     ...     ...     ...     ...  ...   
259  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   
260  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   
261  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   
262  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   
263  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   

          2008       2009       2010       2011       2012       2013  \
0    56.217000  56.579000  56.941000  57.302000  57.636000  57.942000   
1    76.054000  75.687000  75.311000  74.926000  74.532000  74.129000   
2    61.491000  60.701000  59.903000  59.100000  58.301000  57.510000   
3    50.009000  48.924000  47.837000  46.753000  45.670000  44.617000   
4    11.133000  11.648000  12.183000  12.740000  13.292000  13.835000   
..         ...        ...        ...        ...        ...        ...   
259  97.794670  98.377060   0.000000  99.430435  99.773651  99.942505   
260  60.386539  61.751400  63.134281  64.532181  65.942101  75.600000   
261  81.900000  82.700000  82.900000  84.700000  85.300000  85.400000   
262  22.714685  23.235649  22.000000  24.328644  24.894667  25.469702   
263  35.457310  43.369082  36.063057  36.900000  36.728878  37.076813   

          2014        2015        2016    2017  
0    58.221000   58.472000   58.696000  58.893  
1    73.718000   73.297000   72.868000  72.430  
2    56.726000   55.950000   55.181000  54.422  
3    43.591000   42.593000   41.624000  40.684  
4    14.367000   14.885000   15.388000  15.873  
..         ...         ...         ...     ...  
259  99.993134  100.000000  100.000000     NaN  
260  66.100000   70.213905   71.642349     NaN  
261  86.000000   85.500000   84.200000     NaN  
262  27.900000   31.100000   27.219337     NaN  
263  32.300000   33.700000   38.145138     NaN  

[528 rows x 62 columns]

# Exercise 2 (Challenge)

This exercise is more challenging.

Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [ ]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value

# TODO: drop any columns from the data frames that aren't needed

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year

# TODO: sort the results by country and then by year

In [18]:
years = list(range(1960,2018))
years = [str(year) for year in years]

In [21]:
df_rural_melted = pd.melt(df_rural, id_vars=['Country Name', 'Country Code'], value_vars=years).\
                        rename(columns={'variable':'Year',
                                        'value':'Rural_Value'})

In [23]:
df_electricity_melted = pd.melt(df_electricity, id_vars=['Country Name', 'Country Code'], value_vars=years).\
                        rename(columns={'variable':'Year',
                                        'value':'Electricity_Value'})

In [ ]:
pd.merge()

In [35]:
df_combined = df_rural_melted\
                .merge(df_electricity_melted[['Country Name','Year','Electricity_Value']], 
                                left_on=['Country Name','Year'], right_on=['Country Name','Year'])

In [36]:
df_combined.head()

Country Name Country Code  Year  Rural_Value  Electricity_Value
0        Aruba          ABW  1960       49.224                NaN
1  Afghanistan          AFG  1960       91.779                NaN
2       Angola          AGO  1960       89.565                NaN
3      Albania          ALB  1960       69.295                NaN
4      Andorra          AND  1960       41.550                NaN